lib/amqp-common/auth/aad.ts

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+import { ApplicationTokenCredentials, DeviceTokenCredentials, UserTokenCredentials, MSITokenCredentials } from "ms-rest-azure";
+import { TokenInfo, TokenType, TokenProvider } from "./token";
+import * as Constants from "../util/constants";
+
+/**
+ * Defines the AAD (Azure ActiveDirectory) TokenProvider.
+ * @class AadTokenProvider
+ */
+export class AadTokenProvider implements TokenProvider {
+  /**
+   * @property {(ApplicationTokenCredentials | UserTokenCredentials | DeviceTokenCredentials | MSITokenCredentials)} credentials - The credentials object after successful authentication with AAD.
+   */
+  credentials: ApplicationTokenCredentials | UserTokenCredentials | DeviceTokenCredentials | MSITokenCredentials;
+  /**
+   * @property {number} tokenRenewalMarginInSeconds - The number of seconds within which it is
+   * good to renew the token. A constant set to 270 seconds (4.5 minutes). Adal has a set window of 5 minutes
+   * when it refreshes the token from its token cache.
+   */
+  readonly tokenRenewalMarginInSeconds: number = 270;
```

Comment:
i'm probably misunderstanding but does this mean the token gets renewed every 4.5 minutes or every 55.5 minutes? an interval of 4.5 minutes when the token duration is 1 hour seems overzealous.

Reply:
Oh this is the margin (remaining time in 1 hour). This is used for AAD Token Credentials. If I try to renew the token before 55 minutes then ADAL will give me the stale token from its token cache. Hence I make sure that the call to get the token is made within adal's token renewal window, so I get a valid token for the next hour.